src/Features/LanguageServer/Protocol/Handler/AbstractRequestHandlerProvider.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Defines a provider to create instances of <see cref="IRequestHandler"/>.
+    /// New handler instances are created for each LSP server and re-created whenever the 
+    /// server restarts.
+    /// 
+    /// Each <see cref="AbstractRequestHandlerProvider"/> can create multiple <see cref="IRequestHandler"/>
+    /// instances in order to share state between different LSP methods.
+    /// E.g. completion requests can share a cache with completion resolve requests for the same LSP server.
+    /// </summary>
+    internal abstract class AbstractRequestHandlerProvider
+    {
+        /// <summary>
+        /// Instantiates new handler instances and returns the instance
+        /// with the LSP method metadata.
+        /// </summary>
+        public IEnumerable<(IRequestHandler Handler, ILspMethodMetadata Metadata)> CreateRequestHandlers()
```

Comment:
I decided to go the route of having the provider be able to create multiple request handlers for the following reasons:
1.  We need to share state between different LSP methods (e.g. completion and completionresolve).  With this method I can instantiate the cache and give the same one to both handlers (but ofc different servers get a separate call to CreateRequestHandlers)
2. I attempted to unify all LSP methods using the same cache into the same request handler instance.  However that doesn't help much. Multiple custom export attributes (e.g. [ExportLspProvider(m1, mutates:false), ExportLspProvider(m2, mutates:true)] are not supported.  
3. With the same handler instance for multiple LSP methods it's possible to get method signature conflicts if the LSP methods define the same input+output params.

As such, I decided to stick with 1 LSP method per request handler, and just give them all the same shared cache instances.